arbitrator/wasm-libraries/user-host/src/host.rs

```diff
@@ -1,331 +1,690 @@
+        if $env.evm_data().tracing {
+            //let start_ink = $env.start_ink;
+            let end_ink = $env.ink_ready()?;
+            let mut args = vec![];
+            $(args.extend($args);)*
+            let mut outs = vec![];
+            $(outs.extend($outs);)*
+            $env.trace($name, &args, &outs, end_ink);
+        }
+        Ok($ret)
+    }};
+    ($name:expr, $env:expr, [$($args:expr),+], $outs:expr) => {{
+        trace!($name, $env, [$($args),+], $outs, ())
+    }};
+    ($name:expr, $env:expr, $args:expr, $outs:expr) => {{
+        trace!($name, $env, $args, $outs, ())
+    }};
+    ($name:expr, $env:expr, [$($args:expr),+], $outs:expr, $ret:expr) => {
+        trace!($name, $env, [$($args),+], [$outs], $ret)
+    };
+    ($name:expr, $env:expr, $args:expr, $outs:expr, $ret:expr) => {
+        trace!($name, $env, [$args], [$outs], $ret)
+    };
 }
+type Address = Bytes20;
+type Wei = Bytes32;
 
-#[no_mangle]
-pub unsafe extern "C" fn user_host__storage_load_bytes32(key: usize, dest: usize) {
-    let program = Program::start(2 * PTR_INK + EVM_API_INK);
-    let key = wavm::read_bytes32(key);
+#[allow(clippy::too_many_arguments)]
+pub trait UserHost: GasMeteredMachine {
+    type Err: From<OutOfInkError> + From<Self::MemoryErr> + From<eyre::ErrReport>;
+    type MemoryErr;
+    type A: EvmApi;
+
+    fn args(&self) -> &[u8];
+    fn outs(&mut self) -> &mut Vec<u8>;
+
+    fn evm_api(&mut self) -> &mut Self::A;
+    fn evm_data(&self) -> &EvmData;
+    fn evm_return_data_len(&mut self) -> &mut u32;
+
+    fn read_bytes20(&self, ptr: u32) -> Result<Bytes20, Self::MemoryErr>;
+    fn read_bytes32(&self, ptr: u32) -> Result<Bytes32, Self::MemoryErr>;
+    fn read_slice(&self, ptr: u32, len: u32) -> Result<Vec<u8>, Self::MemoryErr>;
+
+    fn write_u32(&mut self, ptr: u32, x: u32) -> Result<(), Self::MemoryErr>;
+    fn write_bytes20(&self, ptr: u32, src: Bytes20) -> Result<(), Self::MemoryErr>;
+    fn write_bytes32(&self, ptr: u32, src: Bytes32) -> Result<(), Self::MemoryErr>;
+    fn write_slice(&self, ptr: u32, src: &[u8]) -> Result<(), Self::MemoryErr>;
+
+    fn say<D: Display>(&self, text: D);
+    fn trace(&self, name: &str, args: &[u8], outs: &[u8], end_ink: u64);
+
+    fn read_args(&mut self, ptr: u32) -> Result<(), Self::Err> {
+        self.buy_ink(HOSTIO_INK)?;
+        self.pay_for_write(self.args().len() as u64)?;
+        self.write_slice(ptr, self.args())?;
+        trace!("read_args", self, &[], self.args())
+    }
+
+    fn write_result(&mut self, ptr: u32, len: u32) -> Result<(), Self::Err> {
+        self.buy_ink(HOSTIO_INK)?;
+        self.pay_for_read(len.into())?;
+        *self.outs() = self.read_slice(ptr, len)?;
+        trace!("write_result", self, &*self.outs(), &[])
+    }
+
+    fn storage_load_bytes32(&mut self, key: u32, dest: u32) -> Result<(), Self::Err> {
+        self.buy_ink(HOSTIO_INK + 2 * PTR_INK + EVM_API_INK)?;
```

Comment:
Comment could be helpful to explain the 2 magic number

Reply:
The `2` comes from the number of pointer ops. There's a `read_bytes32` and a `write_bytes32`, so we charge for two `PTR_INK`'s